### ISSUE TITLE
arc deadlock

### ISSUE THREAD

**Issue by alexanderhaensch** [zfsonlinux/zfs#3958]
We had a problem during a large zfs send operation with a lot of load in parallel.
After a period with slow performance this happened:
https://gist.github.com/alexanderhaensch/93bb037ed3f2f85f9ebd
a problem with l2arc_feed? Our l2arc is very large so i can not imagine that is getting full.

A few days later thing got more intense : https://gist.github.com/alexanderhaensch/873fa318e44bc7a810b0
A reboot that did not allow clean umount anymore (60minutes wait time)

 ------------[ cut here ]------------
 WARNING: CPU: 3 PID: 31720 at fs/dcache.c:1264 umount_check+0x74/0x80()
 Modules linked in: arc4 ecb md4 cifs binfmt_misc iscsi_target_mod target_core_pscsi target_core_file target_core_iblock target_core_mod veth bonding bridge stp llc x86_pkg_temp_thermal coretemp kvm_intel kvm ixgbe crc32c_intel aesni_intel lrw glue_helper ablk_helper cryptd pcspkr joydev mdio microcode isci i2c_i801 ioatdma acpi_cpufreq lpc_ich processor dca mfd_core thermal_sys wmi button nfsd auth_rpcgss oid_registry nfs_acl ipv6 xts gf128mul aes_x86_64 cbc sha512_generic sha1_generic libiscsi scsi_transport_iscsi tg3 libphy e1000 zfs(PO) zunicode(PO) zcommon(PO) znvpair(PO) zavl(PO) spl(O) nfs lockd sunrpc fscache dm_snapshot dm_bufio dm_crypt dm_mirror dm_region_hash dm_log hid_sunplus hid_sony led_class hid_samsung hid_pl hid_petalynx hid_gyration sl811_hcd usbhid xhci_hcd ohci_pci
  ohci_hcd uhci_hcd usb_storage aic94xx lpfc megaraid_sas megaraid_mbox megaraid_mm megaraid aacraid sx8 DAC960 cciss 3w_9xxx 3w_xxxx mptsas mptfc scsi_transport_fc scsi_tgt mptspi mptscsih mptbase atp870u dc395x qla1280 imm parport dmx3191d sym53c8xx gdth advansys initio BusLogic arcmsr aic7xxx aic79xx scsi_transport_spi sg pdc_adma sata_inic162x sata_mv ata_piix sata_qstor sata_vsc sata_uli sata_sis sata_sx4 sata_nv sata_via sata_svw sata_sil24 sata_sil sata_promise pata_sl82c105 pata_cs5530 pata_cs5520 pata_via pata_jmicron pata_marvell pata_sis pata_netcell pata_sc1200 pata_pdc202xx_old pata_triflex pata_atiixp pata_opti pata_amd pata_ali pata_it8213 pata_pcmcia pcmcia pcmcia_core pata_ns87415 pata_ns87410 pata_serverworks pata_artop pata_it821x pata_optidma pata_hpt3x2n pata_hpt3x3
  pata_hpt37x pata_hpt366 pata_cmd64x pata_efar pata_rz1000 pata_sil680 pata_radisys pata_pdc2027x pata_mpiix ehci_pci ehci_hcd libsas ahci libahci usbcore libata usb_common
 CPU: 3 PID: 31720 Comm: umount Tainted: P           O 3.14.51-hardened #1
 Hardware name: Supermicro X9DR3-F/X9DR3-F, BIOS 3.0a 07/31/2013
  0000000000000009 ffff8806956efcd0 ffffffff815f25a5 0000000000000000
  ffff8806956efd08 ffffffff81042743 ffff881ffb553140 ffff881020812800
  ffffffffa0aa8180 ffff880ffbd7a690 ffff880ffbd7a660 ffff8806956efd18
 Call Trace:
  [<ffffffff815f25a5>] dump_stack+0x45/0x56
  [<ffffffff81042743>] warn_slowpath_common+0x73/0x90
  [<ffffffffa0aa8180>] ? _fini+0xd697/0x7b2a3 [zfs]
  [<ffffffff81042815>] warn_slowpath_null+0x15/0x20
  [<ffffffff81131274>] umount_check+0x74/0x80
  [<ffffffff8112f3b5>] d_walk+0x65/0x2b0
  [<ffffffffa0aa8180>] ? _fini+0xd697/0x7b2a3 [zfs]
  [<ffffffff81131200>] ? dget_parent+0x70/0x70
  [<ffffffffa0aa8180>] ? _fini+0xd697/0x7b2a3 [zfs]
  [<ffffffff811310e1>] do_one_tree+0x21/0x40
  [<ffffffff81131aba>] shrink_dcache_for_umount+0x2a/0x90
  [<ffffffff8111b04c>] generic_shutdown_super+0x1c/0xf0
  [<ffffffff8111b36d>] kill_anon_super+0xd/0x20
  [<ffffffffa0a90015>] zpl_kill_sb+0x15/0x20 [zfs]
  [<ffffffff8111b69c>] deactivate_locked_super+0x3c/0x60
  [<ffffffff8111bc23>] deactivate_super+0x43/0x60
  [<ffffffff81138082>] mntput_no_expire+0xa2/0x130
  [<ffffffff8113812f>] mntput+0x1f/0x30
  [<ffffffff811381fa>] namespace_unlock+0xba/0xd0
  [<ffffffff81139367>] SyS_umount+0x207/0x3f0
  [<ffffffff815f897e>] system_call_fastpath+0x16/0x1b
 ---[ end trace c0f88b737093ecb2 ]---

0.6.4.2 with 3.14.53 kernel.
128GB RAM

**Comment by Sachiru** [comment]
Could you try upgrading to 0.6.5.3?

There are 10 kinds of people in the world; those who can read binary and
those who can't.

**Comment by alexanderhaensch** [comment]
Yes already done.

**Comment by tuxoko** [comment]
Hi @alexanderhaensch 
The first two lockup is the same the resilver one you posted.
The umount_check seems to be snapshot automount issue that should be fixed in 0.6.5.*

**Comment by behlendorf** [comment]
@alexanderhaensch is this still an issue in the version of ZFS your using?

**Comment by alexanderhaensch** [comment]
This was not happening any more with tag 0.6.5.3
